src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -94,6 +95,35 @@ variables {G}
 @[pattern] abbreviation cons' (u v w : V) (h : G.adj u v) (p : G.walk v w) : G.walk u w :=
 walk.cons h p
 
+/-- Change the endpoints of a walk using equalities. This is helpful for relaxing
+definitional equality constraints and to be able to state otherwise difficult-to-state
+lemmas. While this is a simple wrapper around `eq.rec`, it gives a canonical way to write it.
+
+The simp-normal form is for the `copy` to be pushed outward. That way calculations can
+occur within the "copy context." -/
+protected def copy {u v u' v'} (p : G.walk u v) (hu : u = u') (hv : v = v') : G.walk u' v' :=
+eq.rec (eq.rec p hv) hu
+
+@[simp] lemma copy_rfl_rfl {u v} (p : G.walk u v) (hu : u = u) (hv : v = v) :
+  p.copy hu hv = p := rfl
```

Comment:
I don't think there's quite a consensus on what should be done with tautologies within arguments on lemmas. I think they can mess up `simp` or `rw` if the proofs are inlined, but this should be tested.

Reply:
Do you know of any other lemmas with tautologies within arguments? I had thought we did have consensus on this already, and it's already been tested.

Reply:
There's [`finset.max'_singleton`](https://leanprover-community.github.io/mathlib_docs/data/finset/lattice.html#finset.max'_singleton) which inlines a tautology, but there's also [`well_founded.not_lt_min`](https://leanprover-community.github.io/mathlib_docs/order/well_founded.html#well_founded.not_lt_min) which takes an `h : set.nonempty s` argument despite being inferrable from a separate `x ∈ s` argument. That's the only other examples I know.

Reply:
Sorry, I meant more specifically of examples of the latter - other than `well_founded.not_lt_min` I don't know of any other such examples. Some more examples of the consensus are around `finset.max'_singleton` - I count about 30 similar examples in that file alone where a proof is given in the statement, rather than as an assumption. There are further examples for lists: `list.last_cons` and many of the lemmas below it, for instance. To the best of my knowledge, the pattern is almost always to prove things rather than assume them

Reply:
I had thought there were some cases where `simp` or `rw` worked more reliably with generalized proofs like this, but I just ran through a few configurations and it always appears to succeed, so I'll just switch to `rfl` here.